### PR TITLE
QuickTalk: data-fetching with SWR

### PR DESCRIPTION
**Pull request by eatyourgreens**
This converts all the `QuickTalk` data fetching functions to use SWR (stale-while-revalidate.) New comments are parsed out of the POST response and appended to the `comments` array, so that they appear without refreshing the whole list and losing your place.

~~Labelled as draft until comments are refreshed after you post a new comment.~~

## Package
lib-classifier

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
- What user actions should my reviewer step through to review this PR?
- Which storybook stories should be reviewed?
- Are there plans for follow up PR’s to further fix this bug or develop this feature?

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Updated to push new comments onto the end of the `comments` array, which seems to work but might not refresh authors or author roles.

**Comment by eatyourgreens** [comment]
New comments are working now so I'm going to remove draft status.

File: packages/lib-classifier/src/components/Classifier/components/QuickTalk/QuickTalkContainer.js [hunk 1]
```diff
@@ -144,21 +94,21 @@ function QuickTalkContainer ({
...
 
         setPostCommentStatus(asyncStates.success)
         setPostCommentStatusMessage('')
-        fetchComments()
 
       } else {  // Create a new discussion
 
-        const response = await postTalkDiscussion (text, discussionTitle, subject, defaultBoard, user, authorization)
-        if (!response?.ok) throw new Error(t('QuickTalk.errors.failPostDiscussion'))
+        const discussion = await postTalkDiscussion (text, discussionTitle, subject, defaultBoard, user, authorization)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/QuickTalk/QuickTalkContainer.js:106]: Some odd whitespace here, when the function is invoked. It’s ignored but looks a bit strange when you read it.

**Review by shaunanoordin — APPROVED** [review]
## PR Overview

Package: lib-classifier

This PR is a refactor & update of the QuickTalk component.

- Notable change: user resource now managed by `usePanoptesUser()`
- Notable change: `swr` is now used to fetch comments (and refresh whenever the subject changes)
- New behaviour: when a new comment is successfully posted, that comment is appended to the current comments, instead of re-fetching the comments.

Code read looks good, and testing looks good too.

### Testing

Tested on `app-project` on localhost with macOS + Chrome 102

- Testing URL: https://local.zooniverse.org:3000/projects/darkeshard/test-project-2022/classify/workflow/3581?env=staging - Classifier page for a project with QuickTalk enabled, with several Subjects that have comments.
- I can view the QuickTalk button.
- I can open the QuickTalk panel and view all comments.
- If logged in, I can post a comment.

### Status

LGTM 👍

File: packages/lib-classifier/src/components/Classifier/components/QuickTalk/QuickTalkContainer.js [hunk 2]
```diff
@@ -44,76 +46,47 @@ function storeMapper (store) {
...
-  function onMount () {
-    fetchComments()
-    checkUser()
-    return () => {}
-  }
+  let author_ids = comments?.map(comment => comment.user_id)
+  author_ids = author_ids?.filter((id, i) => author_ids.indexOf(id) === i)  // Remove duplicates
 
-  useEffect(onMount, [subject])
+  const authors = {}
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/QuickTalk/QuickTalkContainer.js:71]: I found the new React docs useful when thinking about [values that are derived from props](https://beta.reactjs.org/apis/usestate#storing-information-from-previous-renders). Under the hood, I think `useSWR` acts like a state hook.

**Pull request merged**